### PR TITLE
Check `nodetool upgradesstables` works after all nodes upgraded

### PR DESCRIPTION
**Pull request by fruch**
## PR pre-checks (self review)

- [x] I followed [KISS principle](https://en.wikipedia.org/wiki/KISS_principle)
- [x] I gave variables/functions meaningful self-explanatory names
- [x] I didn't leave commented-out/debugging code
- [x] I didn't copy-paste code
- [x] I added the relevant `backport` labels
- [ ] ~~New configuration option are added and documented (in `sdcm/sct_config.py`)~~
- [ ] ~~I have added tests to cover my changes (Infrastructure only - under `unit-test/` folder)~~
- [x] All new and existing unit tests passed (`hydra unit-tests`)
- [ ] ~~I have updated the Readme/doc folder accordingly (if needed)~~

### CODE REVIEW

File: upgrade_test.py [hunk 1]
```diff
@@ -355,6 +387,10 @@ def test_rolling_upgrade(self):
 
         self.verify_stress_thread(entire_write_cs_thread_pool)
 
+        # run 'nodetool upgradesstables' on all nodes and check/wait for all file to be upgraded
```
Comment (amoskong) [inline upgrade_test.py]: We already executed `nodetool upgradesstables` in each upgrade/rollback. Why do we repeatedly upgrade sstables here?

File: upgrade_test.py [hunk 2]
```diff
@@ -198,12 +202,40 @@ def rollback_node(self, node):
...
+
+            sstable_versions = list(
+                set([sstable_version_regex.search(f).group(1) for f in all_sstable_files if sstable_version_regex.search(f)]))
+
+            assert len(sstable_versions) == 1, "expected all table format to be the same found {}".format(sstable_versions)
+            assert sstable_versions[0] == 'mc', "expected to format version to be '{}', found '{}'".format(
+                self.expected_sstable_format_version, sstable_versions[0])
+
+        try:
+            self.debug("Starting to wait for upgardesstables to finish")
```
Comment (amoskong) [inline upgrade_test.py]: If upgradesstables command isn't available, we don't need to wait.

Comment (fruch) [reply upgrade_test.py]: you got a point here :)
I'll fix it

File: upgrade_test.py (hunk 1 above)
Comment (fruch) [reply upgrade_test.py]: me and @roydahan set over the logs, and seen that doing those upgrades while not all the cluster is upgraded, is no-op, i.e. there's no agreement yet in the cluster on the sstable format.

hence we decided to add this stage, to upgrade everything, and check everything is upgraded.

any idea on figuring what is the latest format version ?  that we expect ? maybe grepping the logs for the features ?

Comment (amoskong) [reply upgrade_test.py]: Can we remove upgradesstables in each upgrade or rollback, and only upgrade them in the end?

The easy way to identify the format is checking sstable file name.
- la-...  old format
- mc-... new sst3 format

Comment (fruch) [reply upgrade_test.py]: I think it's good to cover the behavior of using this tool in cluster not fully upgraded, to make sure we don't leave the cluster in a state we can't rollback.

as for the filenames I know that, question is what is the highest version scylla support and if we have a way to query it without hardcoding it in the code, as I'm doing now

File: upgrade_test.py (hunk 2 above)
Comment (fruch) [reply upgrade_test.py]: fixed that

File: upgrade_test.py (hunk 1 above)
Comment (amoskong) [reply upgrade_test.py]: sst3 format is supported from opensource 3.0, enterprise 2019.1

Comment (amoskong) [reply upgrade_test.py]: A simple way is checking `test_sst3` parameter.
We only execute upgradesstables when we sst3 format is enabled.

Comment (fruch) [reply upgrade_test.py]: a question what's sst stands for ?

and still I was thinking on a better way to achieve get what is configured, without prior knowledge of which version support what

Comment (fruch) [reply upgrade_test.py]: @roydahan, 
what's you take on leaving only the `nodetool upgradesstables` after the whole cluster is upgraded ? or we are good with it as it is ?

**Comment by fruch** [comment]
@roydahan one more observation I've found, the command does wait until it's finished.
and in our upgrade test it took less then 2min to finish on the whole cluster

https://cloudius-jenkins-test.s3.amazonaws.com/2a894258-550c-409c-8d6b-10cb814709f4/job-2019-06-12T16.44-0bb235b.zip

File: upgrade_test.py (hunk 1 above)
Comment (amoskong) [reply upgrade_test.py]: > a question what's sst stands for ?

sst -> SSTable
 
> and still I was thinking on a better way to achieve get what is configured, without prior knowledge of which version support what

Agree with you.

Comment (fruch) [reply upgrade_test.py]: see my last commit, trying to address this

**Comment by fruch** [comment]
@amoskong @roydahan, can you guys re-review this one ?

you can see a log of it running in here:
https://jenkins.scylladb.com/view/staging/job/scylla-staging/job/empty/job/rolling-upgrade/job/rolling-upgrade-centos7/9/

File: upgrade_test.py [hunk 3]
```diff
@@ -198,12 +204,69 @@ def rollback_node(self, node):
...
+                result = node.remoter.run("sudo find /var/lib/scylla/data/system -type f ! -path '*snapshots*' | xargs -I{} basename {}")
+                all_sstable_files = result.stdout.splitlines()
+
+                sstable_version_regex = re.compile(r'(\w+)-\d+-(.+)\.(db|txt|sha1|crc32)')
+
+                sstable_versions = list(
+                    set([sstable_version_regex.search(f).group(1) for f in all_sstable_files if sstable_version_regex.search(f)]))
+
+                assert len(sstable_versions) == 1, "expected all table format to be the same found {}".format(sstable_versions)
+                assert sstable_versions[0] == 'mc', "expected to format version to be '{}', found '{}'".format(
```
Comment (bentsi) [inline upgrade_test.py]: `sstable_versions[0] ==  self.expected_sstable_format_version`

**Comment by bentsi** [comment]
@amoskong please review

File: upgrade_test.py (hunk 1 above)
Comment (roydahan) [reply upgrade_test.py]: 1. I think we should leave it as is, IOW test the upgradesstables on each step.
We aren't testing only happy path, we would like to test things that customers may do (like what me and Israel thought).

2. I'm against of using the yaml parameter test_sst3. we should automatically detect it.

File: upgrade_test.py [hunk 4]
```diff
@@ -198,12 +204,69 @@ def rollback_node(self, node):
...
+            queue.put(upgradesstables_available)
+            queue.task_done()
+
+    def get_highest_supported_sstable_version(self):
+        """
+        find the highest sstable format version supported in the cluster
+
+        :return:
+        """
+        sstable_format_regex = re.compile(r'Feature (.*)_SSTABLE_FORMAT is enabled')
```
Comment (roydahan) [inline upgrade_test.py]: Good

**Pull request merged**